src/System.Linq/tests/FirstTests.cs

```diff
@@ -0,0 +1,187 @@
+                    where x > Int32.MinValue
+                    select x;
+
+            Assert.Equal(q.First(), q.First());
+        }
+
+        [Fact]
+        public void SameResultsRepeatCallsStringQuery()
+        {
+            var q = from x in new[] { "!@#$%^", "C", "AAA", "", "Calling Twice", "SoS", String.Empty }
+                    where !String.IsNullOrEmpty(x)
+                    select x;
+
+            Assert.Equal(q.First(), q.First());
+        }
+
+        public void TestEmptyIList<T>()
+        {
+            T[] source = { };
+            
+            Assert.NotNull(source as IList<T>);
+            
+            Assert.Throws<InvalidOperationException>(() => source.First());
+        }
+
+        [Fact]
+        public void EmptyIListT()
+        {
+            TestEmptyIList<int>();
+            TestEmptyIList<string>();
+            TestEmptyIList<DateTime>();
+            TestEmptyIList<FirstTests>();
+        }
+
+        [Fact]
+        public void IListTOneElement()
+        {
+            int[] source = { 5 };
+            int expected = 5;
+
+            Assert.NotNull(source as IList<int>);
+            
+            Assert.Equal(expected, source.First());
+        }
+
+        [Fact]
+        public void IListTManyELementsFirstIsDefault()
+        {
+            int?[] source = { null, -10, 2, 4, 3, 0, 2 };
+            int? expected = null;
+
+            Assert.IsAssignableFrom<IList<int?>>(source);
+            
+            Assert.Equal(expected, source.First());
+        }
+
+        [Fact]
+        public void IListTManyELementsFirstIsNotDefault()
+        {
+            int?[] source = { 19, null, -10, 2, 4, 3, 0, 2 };
+            int? expected = 19;
+
+            Assert.IsAssignableFrom<IList<int?>>(source);
+            
+            Assert.Equal(expected, source.First());
+        }
+
+        private static IEnumerable<T> EmptySource<T>()
+        {
+            yield break;
+        }
+
+        private static void TestEmptyNotIList<T>()
+        {
+            var source = EmptySource<T>();
+
+            Assert.Null(source as IList<T>);
+            
+            Assert.Throws<InvalidOperationException>(() => source.First());
+        }
+
+        [Fact]
+        public void EmptyNotIListT()
+        {
+            TestEmptyNotIList<int>();
+            TestEmptyNotIList<string>();
+            TestEmptyNotIList<DateTime>();
+            TestEmptyNotIList<FirstTests>();
+        }
+
+        [Fact]
+        public void OneElementNotIListT()
+        {
+            IEnumerable<int> source = NumList(-5, 1);
+            int expected = -5;
+
+            Assert.Null(source as IList<int>);
+            
+            Assert.Equal(expected, source.First());
+        }
+
+        [Fact]
+        public void ManyElementsNotIListT()
+        {
+            IEnumerable<int> source = NumList(3, 10);
+            int expected = 3;
+
+            Assert.Null(source as IList<int>);
```

Comment:
These asserts about the results of NumList are really asserting things about test code rather than about source code.  If the goal is to point out that the test depends on this not being an IList, should we just move the assert into NumList itself; that would save these asserts at the call sites, and it would highlight to anyone in the future trying to change NumList that there are dependencies on this condition.